Sources/NIOOpenSSL/ByteBufferBIO.swift

```diff
@@ -0,0 +1,289 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import CNIOOpenSSL
+
+
+/// The OpenSSL entry point to write to the `ByteBufferBIO`. This thunk unwraps the user data
+/// and then passes the call on to the specific BIO reference.
+internal func openSSLBIOWriteFunc(bio: UnsafeMutablePointer<BIO>?, buf: UnsafePointer<CChar>?, len: CInt) -> CInt {
+    guard let concreteBIO = bio, let concreteBuf = buf else {
+        preconditionFailure("Invalid pointers in openSSLBIOWriteFunc: bio: \(String(describing: bio)) buf: \(String(describing: buf))")
+    }
+
+    // This unwrap may fail if the user has dropped the ref to the ByteBufferBIO but still has
+    // a ref to the other pointer. Sigh heavily and just fail.
+    guard let userPtr = CNIOOpenSSL_BIO_get_data(concreteBIO) else {
+        return -1
+    }
+
+    // Begin by clearing retry flags. We do this at all OpenSSL entry points.
+    CNIOOpenSSL_BIO_clear_retry_flags(concreteBIO)
+
+    // In the event a write of 0 bytes has been asked for, just return early, don't bother with the other work.
+    guard len > 0 else {
+        return 0
+    }
+
+    let swiftBIO: ByteBufferBIO = Unmanaged.fromOpaque(userPtr).takeUnretainedValue()
+    let bufferPtr = UnsafeBufferPointer(start: concreteBuf, count: Int(len))
+    return swiftBIO.sslWrite(buffer: bufferPtr)
+}
+
+/// The OpenSSL entry point to read from the `ByteBufferBIO`. This thunk unwraps the user data
+/// and then passes the call on to the specific BIO reference.
+internal func openSSLBIOReadFunc(bio: UnsafeMutablePointer<BIO>?, buf: UnsafeMutablePointer<CChar>?, len: CInt) -> CInt {
```

Comment:
same questions as above

Reply:
As discussed in person, we want to use the old values in the body of the `else` block, which requires we bind to new names.